mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java

```diff
@@ -0,0 +1,145 @@
+import static org.apache.james.mailbox.cassandra.table.CassandraQuotaLimit.QUOTA_COMPONENT;
+import static org.apache.james.mailbox.cassandra.table.CassandraQuotaLimit.QUOTA_LIMIT;
+import static org.apache.james.mailbox.cassandra.table.CassandraQuotaLimit.QUOTA_SCOPE;
+import static org.apache.james.mailbox.cassandra.table.CassandraQuotaLimit.QUOTA_TYPE;
+import static org.apache.james.mailbox.cassandra.table.CassandraQuotaLimit.TABLE_NAME;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.core.quota.QuotaComponent;
+import org.apache.james.core.quota.QuotaLimit;
+import org.apache.james.core.quota.QuotaScope;
+import org.apache.james.core.quota.QuotaType;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.api.core.cql.Row;
+import com.datastax.oss.driver.api.querybuilder.delete.Delete;
+import com.datastax.oss.driver.api.querybuilder.insert.Insert;
+import com.datastax.oss.driver.api.querybuilder.select.Select;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class CassandraQuotaLimitDao {
+
+    private final CassandraAsyncExecutor queryExecutor;
+    private final PreparedStatement getQuotaLimitStatement;
+    private final PreparedStatement getQuotaLimitsStatement;
+    private final PreparedStatement setQuotaLimitStatement;
+    private final PreparedStatement deleteQuotaLimitStatement;
+
+    @Inject
+    public CassandraQuotaLimitDao(CqlSession session) {
+        this.queryExecutor = new CassandraAsyncExecutor(session);
+        this.getQuotaLimitStatement = session.prepare(getQuotaLimitStatement().build());
+        this.getQuotaLimitsStatement = session.prepare(getQuotaLimitsStatement().build());
+        this.setQuotaLimitStatement = session.prepare(setQuotaLimitStatement().build());
+        this.deleteQuotaLimitStatement = session.prepare((deleteQuotaLimitStatement().build()));
+    }
+
+    public Mono<QuotaLimit> getQuotaLimit(QuotaLimit.QuotaKey quotaKey) {
+        return queryExecutor.executeSingleRow(getQuotaLimitStatement.bind()
+            .setString(QUOTA_COMPONENT, quotaKey.getQuotaComponent().getValue())
+            .setString(QUOTA_SCOPE, quotaKey.getQuotaScope().getValue())
+            .setString(IDENTIFIER, quotaKey.getIdentifier())
+            .setString(QUOTA_TYPE, quotaKey.getQuotaType().getValue()))
+            .map(row -> convertRowToModel(row));
+    }
+
+    public Flux<QuotaLimit> getQuotaLimits(QuotaLimit.QuotaKey quotaKey) {
+        return queryExecutor.executeRows(getQuotaLimitsStatement.bind()
+            .setString(QUOTA_COMPONENT, quotaKey.getQuotaComponent().getValue())
+            .setString(QUOTA_SCOPE, quotaKey.getQuotaScope().getValue())
+            .setString(IDENTIFIER, quotaKey.getIdentifier()))
+            .map(row -> convertRowToModel(row));
+    }
```

Comment:
IMO at this low-level DAO layer, we can accept lower abstraction here (just listing the method params instead of reusing QuotaLimit.QuotaKey).

Especially in this case: the same `QuotaLimit.QuotaKey` param is used for `getQuotaLimit` and `getQuotaLimits` methods seems easily to make higher caller confused. (The caller needs to check the code to see if it can put a `null` `QUOTA_TYPE` or not).

Again we do not tend to use `null` like that. If you want to put a potential null in a method params, use Optional and let the DAO handles the null logic.

Reply:
Also would not it be weird if e.g we call `.getQuotaLimits` with a `QuotaKey` containing a `count` `QuotaType` object and then we receive a size `QuotaType` record?

Reply:
Ok probably for the case getQuotaLimits, I will use component, scope and identifier as method paras